dspy/utils/usage_tracker.py

```diff
@@ -0,0 +1,59 @@
+"""Usage tracking utilities for DSPy."""
+
+from collections import defaultdict
+from contextlib import contextmanager
+from typing import Any
+
+from dspy.dsp.utils.settings import settings
+
+
+class UsageTracker:
+    """Tracks LM usage data within a context."""
+
+    def __init__(self):
+        self.usage_data = defaultdict(list)
```

Comment:
nit: if may be helpful if we document what are the key and value of this dict

Reply:
sg!